examples/hpc/dla-driver/src/lib.rs

```diff
@@ -18,291 +100,493 @@ macro_rules! set_bits {
+        unsafe { ptr::read_volatile((DLA0_ADDR + offset) as *mut u32) }
+    }
 
-pub fn dla_set_kernel_data_bank(bank: usize) {
-    let mut reg = dla_read_reg(DLA_BUF_DATA_BANK);
-    reg = set_bits!(
-        DLA_BUF_DATA_BANK_A_OFFSET,
-        DLA_BUF_DATA_BANK_A_BITMASK,
-        reg,
-        bank
-    );
-    dla_write_reg(DLA_BUF_DATA_BANK, reg);
-}
+    pub fn read(&self, buf: &mut [u8], len: usize, offset: usize) {
+        for i in 0..len {
+            unsafe { buf[i] = ptr::read_volatile((DLA0_ADDR + offset + i) as *mut u8) }
+        }
+    }
 
-pub fn dla_set_input_size(channels: usize, width: usize, height: usize) {
-    let mut reg = 0;
-    reg = set_bits!(
-        DLA_BUF_INPUT_CHANNELS_OFFSET,
-        DLA_BUF_INPUT_CHANNELS_BITMASK,
-        reg,
-        channels - 1
-    );
-    reg = set_bits!(
-        DLA_BUF_INPUT_WIDTH_OFFSET,
-        DLA_BUF_INPUT_WIDTH_BITMASK,
-        reg,
-        width - 1
-    );
-    reg = set_bits!(
-        DLA_BUF_INPUT_HEIGHT_OFFSET,
-        DLA_BUF_INPUT_HEIGHT_BITMASK,
-        reg,
-        height - 1
-    );
-    dla_write_reg(DLA_BUF_INPUT, reg);
-}
+    pub fn write_data_bank(&self, offset: usize, buf: &mut [u8]) {
+        //sprintln!("\nWrite to bank {:#x}, data: {:?}", offset, buf);
+        for (i, b) in buf.iter().enumerate() {
+            unsafe { ptr::write_volatile((MEMORY_BANK_BASE_ADDR + offset + i) as *mut u8, *b) };
+        }
+    }
 
-pub fn dla_set_kernel_size(channels: usize, width: usize, height: usize) {
-    let mut reg = 0;
-    reg = set_bits!(
-        DLA_BUF_KERNEL_0_S_CHANNELS_OFFSET,
-        DLA_BUF_KERNEL_0_S_CHANNELS_BITMASK,
-        reg,
-        channels - 1
-    );
-    reg = set_bits!(
-        DLA_BUF_KERNEL_0_WIDTH_OFFSET,
-        DLA_BUF_KERNEL_0_WIDTH_BITMASK,
-        reg,
-        width - 1
-    );
-    reg = set_bits!(
-        DLA_BUF_KERNEL_0_HEIGHT_OFFSET,
-        DLA_BUF_KERNEL_0_HEIGHT_BITMASK,
-        reg,
-        height - 1
-    );
-    dla_write_reg(DLA_BUF_KERNEL_0, reg);
-}
+    pub fn read_data_bank_offset(&self, offset: usize) -> u128 {
+        // NOTE: this function enforces the 128-bit addressing
+        if cfg!(feature = "vp") {
+            let mut result: u128 = 0;
+            for i in 0..4 {
+                result |= (unsafe {
+                    ptr::read_volatile((MEMORY_BANK_BASE_ADDR + offset + (i * 4)) as *mut u32)
+                } as u128)
+                    << (32 * i)
+            }
+            result
+        } else {
+            unsafe { ptr::read_volatile((MEMORY_BANK_BASE_ADDR + (offset & !0xF)) as *mut u128) }
+        }
+    }
 
-pub fn dla_input_data_ready(ready: bool) {
-    let mut reg = dla_read_reg(DLA_BUF_CTRL);
-    reg = set_bits!(
-        DLA_READ_B_VALID_OFFSET,
-        DLA_READ_B_VALID_BITMASK,
-        reg,
-        ready as usize
-    );
-    dla_write_reg(DLA_BUF_CTRL, reg);
-}
+    pub fn read_data_bank(&self, offset: usize, len: usize) -> Vec<u8> {
```

Comment:
Does this need to be public? It seems to be referenced only by DLA driver functions.

Reply:
There is something suspicious about this API. It seems like it would allow arbitrary memory reads. Would it be possible to constrain its capabilities by using the `Bank` enum as a parameter instead of the `offset`?

Reply:
So the asic DLA can write results to any arbitrary memory address in SDRAM, thus the DLA driver needs to be able to do the same. The method `read_data_bank()` is thus poorly named. I implemented the restriction for this method to only allow memory access for the banks, but in the future dla driver needs to be able to read the result from arbitrary memory locations.